security/crypto/bls12/bls12.go

```diff
@@ -350,75 +343,85 @@ func (bls *bls12Base) Combine(signatures ...hotstuff.QuorumSignature) (combined
 			}
 			g2.Add(&agg, &agg, &sig2.sig)
 		} else {
-			bls.logger.Panicf("cannot combine incompatible signature type %T (expected %T)", sig1, sig2)
+			return nil, fmt.Errorf("cannot combine incompatible signature type %T (expected %T)", sig1, sig2)
 		}
 	}
 	return &AggregateSignature{sig: agg, participants: participants}, nil
 }
 
 // Verify verifies the given quorum signature against the message.
-func (bls *bls12Base) Verify(signature hotstuff.QuorumSignature, message []byte) bool {
+func (bls *bls12Base) Verify(signature hotstuff.QuorumSignature, message []byte) (err error) {
 	s, ok := signature.(*AggregateSignature)
 	if !ok {
-		bls.logger.Panicf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
+		return fmt.Errorf("cannot verify signature of incompatible type %T (expected %T)", signature, s)
 	}
 
 	n := s.Participants().Len()
 
 	if n == 1 {
 		id := firstParticipant(s.Participants())
-		pk, ok := bls.publicKey(id)
-		if !ok {
-			bls.logger.Warnf("Missing public key for ID %d", id)
-			return false
+		pk, err := bls.publicKey(id)
+		if err != nil {
+			return err
+		}
+		if !bls.coreVerify(pk, message, &s.sig, domain) {
+			return fmt.Errorf("core-verify failed")
 		}
-		return bls.coreVerify(pk, message, &s.sig, domain)
 	}
 
 	// else if l > 1:
 	pks := make([]*PublicKey, 0, n)
+	var errs error
 	s.Participants().RangeWhile(func(id hotstuff.ID) bool {
-		pk, ok := bls.publicKey(id)
-		if ok {
-			pks = append(pks, pk)
-			return true
+		pk, err := bls.publicKey(id)
+		if err != nil {
+			errs = errors.Join(err)
+			return false
 		}
-		bls.logger.Warnf("Missing public key for ID %d", id)
-		return false
+		pks = append(pks, pk)
+		return true
 	})
-	if len(pks) != n {
-		return false
+	if errs != nil {
+		return fmt.Errorf("one or more public keys had errors: %v", errs)
+	}
+	if !bls.fastAggregateVerify(pks, message, &s.sig) {
+		return fmt.Errorf("fast-agg-verify failed")
 	}
-	return bls.fastAggregateVerify(pks, message, &s.sig)
+	return nil
 }
 
 // BatchVerify verifies the given quorum signature against the batch of messages.
-func (bls *bls12Base) BatchVerify(signature hotstuff.QuorumSignature, batch map[hotstuff.ID][]byte) bool {
+func (bls *bls12Base) BatchVerify(signature hotstuff.QuorumSignature, batch map[hotstuff.ID][]byte) error {
 	s, ok := signature.(*AggregateSignature)
 	if !ok {
-		bls.logger.Panicf("cannot verify incompatible signature type %T (expected %T)", signature, s)
+		return fmt.Errorf("cannot verify incompatible signature type %T (expected %T)", signature, s)
 	}
 
 	if s.Participants().Len() != len(batch) {
-		return false
+		return fmt.Errorf("expected %d participants", len(batch))
```

Comment:
Not sure about this one, but maybe something like this:
`signature mismatch: %s participants, expected: %s`